### PR TITLE
Add support for PKCS#11 v3.0 SHA3 mechanisms

### PR DESCRIPTION
**Pull request by ifranzki**

### CODE REVIEW

**Comment by ifranzki** [comment]
Hm... Travis: "The job exceeded the maximum log length, and has been terminated."
I need to think about how to shorten the testcase output....

File: usr/lib/soft_stdll/soft_specific.c [hunk 1]
```diff
@@ -73,12 +73,20 @@ static const MECH_LIST_ELEMENT soft_mech_list[] = {
     {CKM_SHA256_RSA_PKCS, {512, 4096, CKF_SIGN | CKF_VERIFY}},
     {CKM_SHA384_RSA_PKCS, {512, 4096, CKF_SIGN | CKF_VERIFY}},
     {CKM_SHA512_RSA_PKCS, {512, 4096, CKF_SIGN | CKF_VERIFY}},
+    {CKM_SHA3_224_RSA_PKCS, {512, 4096, CKF_SIGN|CKF_VERIFY}},
```
Comment (jschmidb) [inline usr/lib/soft_stdll/soft_specific.c:76]: spaces around the | ?
Also below

File: usr/lib/ica_s390_stdll/ica_specific.c [hunk 2]
```diff
@@ -4609,18 +4641,30 @@ static const REF_MECH_LIST_ELEMENT ref_mech_list[] = {
     {SHA512_256, CKM_SHA512_256_HMAC_GENERAL, {128, 2048, CKF_SIGN | CKF_VERIFY}},
 #endif
 #ifdef SHA3_224
+    {SHA3_224, CKM_SHA3_224, {0, 0, CKF_HW|CKF_DIGEST}},
```
Comment (jschmidb) [inline usr/lib/ica_s390_stdll/ica_specific.c:4644]: spaces around the | ?

File: testcases/crypto/rsa.h [hunk 3]
```diff
@@ -186,14 +186,120 @@ struct RSA_GENERATED_TEST_VECTOR rsa_oaep_generated_tv[] = {
...
+        .modbits = 2048,
+        .publ_exp_len = 3,
+        .publ_exp = {0x01, 0x00, 0x01},
+        .inputlen = 16,
+        .oaep_params = {CKM_SHA3_224, CKG_MGF1_SHA3_224, CKZ_DATA_SPECIFIED,
+                        "abcdefghijkl", 12},
+        .keylen = 32,
+        .keytype = {CKM_AES_KEY_GEN, 0, 0}
+    },
+    {       // #16
```
Comment (jschmidb) [inline testcases/crypto/rsa.h:209]: Indentation of comment?

File: testcases/crypto/rsa.h [hunk 4]
```diff
@@ -457,7 +563,127 @@ struct RSA_GENERATED_TEST_VECTOR sha512_rsa_pss_generated_tv[] = {
     }
 };
 
-#define NUM_OF_GENERATED_PSS_TESTSUITES 6
+struct RSA_GENERATED_TEST_VECTOR sha3_224_rsa_pss_generated_tv[] = {
+    {                           // #0
```
Comment (jschmidb) [inline testcases/crypto/rsa.h:567]: indentation of comments?

File: usr/lib/soft_stdll/soft_specific.c (hunk 1 above)
Comment (ifranzki) [reply usr/lib/soft_stdll/soft_specific.c:76]: fixed (several paces)

File: usr/lib/ica_s390_stdll/ica_specific.c (hunk 2 above)
Comment (ifranzki) [reply usr/lib/ica_s390_stdll/ica_specific.c:4644]: fixed (several paces)

File: testcases/crypto/rsa.h (hunk 3 above)
Comment (ifranzki) [reply testcases/crypto/rsa.h:209]: It is a mess all over the place, I am just fixing the newly inserted lines in this PR.

**Comment by ifranzki** [comment]
> Hm... Travis: "The job exceeded the maximum log length, and has been terminated." I need to think about how to shorten the testcase output....

With commit https://github.com/opencryptoki/opencryptoki/pull/777/commits/f8dcb4bb0e5df2576e6e5ab9515a55269a7eccbe travis now succeeds. It will no longer show all testcases output messages, only 'interesting' ones to reduce the log size.

**Comment by ifranzki** [comment]
The travis failures on x86 and arm64 are unrelated.

**Comment by ifranzki** [comment]
@jschmidb please take another look.

File: usr/lib/common/key_mgr.c [hunk 5]
```diff
@@ -102,6 +102,97 @@ CK_RV key_mgr_apply_always_sensitive_never_extractable_attrs(
...
+
+    rc = build_attribute(CKA_ALWAYS_SENSITIVE, &flag, sizeof(CK_BBOOL),
+                         &always_sens_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("Failed to build CKA_ALWAYS_SENSITIVE attribute.\n");
+        goto end;
+    }
+
+    /*
+     * If base key has NEVER_EXTRACTABLE = FASE, the new key does too
```
Comment (jschmidb) [inline usr/lib/common/key_mgr.c:144]: FALSE

Comment (ifranzki) [reply usr/lib/common/key_mgr.c]: fixed

**Comment by ifranzki** [comment]
Rebased to current master to get Travis working on Jammy.

File: usr/lib/common/mech_sha.c [hunk 6]
```diff
@@ -524,3 +525,246 @@ CK_RV ckm_generic_secret_key_gen(STDLL_TokData_t *tokdata, TEMPLATE *tmpl)
...
+        /* Supply CKA_VAUE_LEN since this is required for those key types */
+        rc = build_attribute(CKA_VALUE_LEN, (CK_BYTE*)&derived_keylen,
+                             sizeof(derived_keylen), &vallen_attr);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("Failed to build the attribute from CKA_VALUE_LEN, "
+                        "rc=%s.\n", ock_err(rc));
+            goto end;
+        }
+        break;
+    case CKK_DES:
```
Comment (jschmidb) [inline usr/lib/common/mech_sha.c:704]: Are there no weak keys possible for DES2 and DES3? A DES2 key is just the concatenation of two DES keys, so we could check the two keys individually for weakness .... similar to DES3.

File: usr/lib/cca_stdll/cca_specific.c [hunk 7]
```diff
@@ -242,6 +246,14 @@ static const MECH_LIST_ELEMENT cca_mech_list[] = {
     {CKM_AES_CBC, {16, 32, CKF_HW | CKF_ENCRYPT | CKF_DECRYPT}},
     {CKM_AES_CBC_PAD, {16, 32, CKF_HW | CKF_ENCRYPT | CKF_DECRYPT}},
     {CKM_AES_XTS, {32, 64, CKF_ENCRYPT | CKF_DECRYPT}},
+    {CKM_SHA3_512, {0, 0, CKF_HW | CKF_DIGEST}},
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:249]: Hmm, I would change the order here: 224, 256, 384, 512 as always ...

File: testcases/crypto/rsa.h [hunk 8]
```diff
@@ -9250,12 +9604,30 @@ struct PUBLISHED_TEST_SUITE_INFO rsa_encdec_import_test_suites[] = {
...
         .name = "RSA PKCS OAEP (SHA256)",
         .tvcount = 2,
         .tv = rsa_sha256_pkcs_sigver_published_tv,
         .mech = {CKM_RSA_PKCS_OAEP, &oaep_params_sha256, sizeof(oaep_params_sha256)},
     },
+    {
+        .name = "RSA PKCS OAEP (SHA384)",
+        .tvcount = 2,
+        .tv = rsa_sha384_pkcs_sigver_published_tv,
+        .mech = {CKM_RSA_PKCS_OAEP, &oaep_params_sha384, sizeof(oaep_params_sha256)},
```
[the remark this answers is not in the document]
Comment (ifranzki) [reply testcases/crypto/rsa.h:9623]: fixed

File: usr/include/pkcs11types.h [hunk 9]
```diff
@@ -950,6 +980,11 @@ typedef CK_ULONG CK_MECHANISM_TYPE;
 #define CKM_ECDSA_SHA256               0x00001044
 #define CKM_ECDSA_SHA384               0x00001045
 #define CKM_ECDSA_SHA512               0x00001046
+/* The following are new for v3.0 */
+#define CKM_ECDSA_SHA3_224             0x00001047
+#define CKM_ECDSA_SHA3_256             0x00001048
+#define CKM_ECDSA_SHA3_384             0x00001049
+#define CKM_ECDSA_SHA3_512             0x0000104a
```
Comment (jschmidb) [inline usr/include/pkcs11types.h:987]: Hm, above you used a capital letter 'A' in the hex value 0x0000039A ... maybe you want to make this consistent

Comment (ifranzki) [reply usr/include/pkcs11types.h]: fixed

**Comment by ifranzki** [comment]
Finally Travis completed without canceling.

**Comment by ifranzki** [comment]
Added 2 more commits to reduce the run time of the testcases. Adding these new SHA variants and all the combined mechanisms with them has increased runtime quite a lot.

**Review by jschmidb — APPROVED** [review]

**Pull request merged**